Update registry to match macOS

JaniruTEC opened the pull request:
This PR continues the efforts made in https://github.com/cryptomator/cryptomator-win/pull/35 and aims to mirror the changes introduced by https://github.com/cryptomator/cryptomator-mac/pull/7 on macOS.
The windows installer will contain the same feature set as [the matching file on macOS](https://github.com/cryptomator/cryptomator-mac/blob/8f36f3dc8f03bfb38c7653a1fed7ec9e7c28fc8a/resources/app/Info.plist) after this PR is merged.
This PR is required to close https://github.com/cryptomator/cryptomator/issues/1089 - see there for general motivation and reasoning.

JaniruTEC committed:
Applied suggestions from code review

Applied suggestions given at https://github.com/cryptomator/cryptomator-mac/pull/7#pullrequestreview-575190899
Mirrors https://github.com/cryptomator/cryptomator-mac/commit/f5b7007e58698f2c0fe254df331905bc10c2ae0f

JaniruTEC committed:
Added comments

See: https://github.com/cryptomator/cryptomator-win/issues/33

JaniruTEC committed:
Added MIME types (content types) and perceived types

Reasoning for MIME type (content type) choices:
macOS uses the same MIME types.

Reasoning for MIME type (content type) placement:
.c9s and .c9r share the same mime type, therefore using the method detailed in (1) is not suitable.
Also this method violates (2), which lead to the decision to use the approach outlined in (3), even if it's far less popular in the general public.

Reasoning for perceived types:
.cryptomator encodes a JSON file.
.c9s and c9r don't match any of the perceived types really well. "system" also warns the user, when they try to open such a file.
See (3) for placement.

References:
(1) https://web.archive.org/web/20210127224743/https://condor.depaul.edu/elliott/420/hw/programs/Windows10_IE_MIME-1.pdf
(2) https://web.archive.org/web/20210127221958/https://docs.microsoft.com/en-us/previous-versions/windows/desktop/legacy/cc144156%28v=vs.85%29#do-not-confuse-the-internet-explorer-mime-database-with-file-associations
(3) https://web.archive.org/web/20210127222435/https://docs.microsoft.com/en-us/windows/win32/shell/fa-file-types#setting-optional-subkeys-and-file-type-extension-attributes
(4) https://web.archive.org/web/20200410143821/https://docs.microsoft.com/en-us/windows/win32/shell/fa-perceivedtypes
(5) https://web.archive.org/web/20201026022647/https://docs.microsoft.com/en-us/windows/win32/api/shtypes/ne-shtypes-perceived

JaniruTEC committed:
Switched HKCR to HKA

See:
https://github.com/cryptomator/cryptomator-win/commit/c60947525917e43213576f30ebdf65ba4714a49a#r46325298
https://web.archive.org/web/20210127225700/https://jrsoftware.org/ishelp/index.php?topic=registrysection
https://web.archive.org/web/20210127225601/https://docs.microsoft.com/en-us/windows/win32/sysinfo/hkey-classes-root-key

overheadhunter commented on resources/innosetup/setup.iss:
```diff
@@ -57,15 +57,31 @@ Name: "webdav"; Description: "WebDAV system configuration"; Types: full; ExtraDi
+; .c9r & .c9s filetypes: Add extension, MIME type, perceived type, description and icon for each
+Root: HKA; Subkey: "Software\Classes\.c9r"; ValueType: string;  ValueName: ""; ValueData: "CryptomatorEncryptedData"; Flags: uninsdeletekey
+Root: HKA; Subkey: "Software\Classes\.c9r"; ValueType: string; ValueName: "Content Type"; ValueData: "application/x-vnd.cryptomator.encrypted-data"
+Root: HKA; Subkey: "Software\Classes\.c9r"; ValueType: string; ValueName: "PerceivedType"; ValueData: "system"
```
What other options do we have for `PerceivedType`? Is there something like `binary`?

JaniruTEC replied:
I don't know about custom types, but currently that's all there is:
> The default perceived types are as follows:
- Folder
- Text
- Image
- Audio
- Video
- Compressed
- Document
- System
- Application
- Gamemedia
- Contacts

See: [Here](https://web.archive.org/web/20200410143821/https://docs.microsoft.com/en-us/windows/win32/shell/fa-perceivedtypes)

overheadhunter replied:
Not exactly a lot of choice, here. Entropywise, `Compressed` would come closest, but would be very misleading if 3rd party applications start feeling the urge to "unzip" a c9r file.

JaniruTEC replied:
Correct. And that's exactly what would happen: The user would receive the suggestion to use a ZIP Application to open the file.

overheadhunter commented on resources/innosetup/setup.iss:
```diff
@@ -57,15 +57,31 @@ Name: "webdav"; Description: "WebDAV system configuration"; Types: full; ExtraDi
+Root: HKA; Subkey: "Software\Classes\.c9r"; ValueType: string; ValueName: "Content Type"; ValueData: "application/x-vnd.cryptomator.encrypted-data"
+Root: HKA; Subkey: "Software\Classes\.c9r"; ValueType: string; ValueName: "PerceivedType"; ValueData: "system"
+
+Root: HKA; Subkey: "Software\Classes\.c9s"; ValueType: string;  ValueName: ""; ValueData: "CryptomatorEncryptedData"; Flags: uninsdeletekey
```
Is there any official documentation confirming that it is allowed and intended to use the same name (`CryptomatorEncryptedData`) multiple times for different file extensions?

JaniruTEC replied:
Yes, indeed. See: [Here](https://web.archive.org/web/20210201153726/https://docs.microsoft.com/de-de/windows/win32/shell/how-to-register-a-file-type-for-a-new-application#step-1)

JaniruTEC committed:
Merge branch 'master' into feature/update-registry

JaniruTEC committed:
Changed rules for key deletion

See for removal of "unisdeletekey" on file extensions: https://web.archive.org/web/20200405143227/https://docs.microsoft.com/en-us/windows/win32/shell/how-to-register-a-file-type-for-a-new-application#step-3

See for addition of "deletkey" on ProgIDs: https://web.archive.org/web/20200405143227/https://docs.microsoft.com/en-us/windows/win32/shell/how-to-register-a-file-type-for-a-new-application#step-2

Also see: https://web.archive.org/web/20210203224603/https://docs.microsoft.com/de-de/windows/win32/shell/fa-file-types#deleting-registry-information-during-uninstallation

JaniruTEC committed:
Added cleanup for temporary key, fixed formatting

JaniruTEC committed:
Switched to versioned ProgIDs

See: https://web.archive.org/web/20210203223928/https://docs.microsoft.com/de-de/windows/win32/shell/fa-progids#using-versioned-programmatic-identifiers

JaniruTEC commented:
A note on HKA vs. HKCR:
In administrative install mode the effect is the same for both Roots: Keys are written to `HKLM/Software/Classes`.
In local user install mode the effects differ: Keys are written to `HKLM/Software/Classes` for HKCR and to `HKCU/Software/Classes` for HKA.
The current setup (with it's deletions, etc.) works fine with administrative install, but may cause problems if we open installations for local users. Since it's not recommended to remove the file extension keys (like `.cryptomator`), I changed the script to only delete the ProgIDs (like `Cryptomator.MasterkeyFile.1`).
If the user now installs Cryptomator version X in local mode and installs version X+1 in admin mode, changes to the file extensions in X (like changing the MIME Type) still will outrank the newer administrative changes (like updating the ProgID version).
As mentioned this will only cause problems in the very specific scenario that the user installs an admin version over a local user version and will only affect the file extension entry, not the matching ProgID.

See [here](https://web.archive.org/web/20210203230929/https://docs.microsoft.com/de-de/windows/win32/sysinfo/hkey-classes-root-key) for info about HKCR and [here](https://web.archive.org/web/20210127225700/https://jrsoftware.org/ishelp/index.php?topic=registrysection) for info about HKA.

overheadhunter commented:
> Since it's not recommended to remove the file extension keys (like .cryptomator), I changed the script to only delete the ProgIDs (like Cryptomator.MasterkeyFile.1).

How does appending `.1` to an arbitrary string protect a key from being deleted?

> If the user now installs Cryptomator version X in local mode and installs version X+1 in admin mode, changes to the file extensions in X (like changing the MIME Type) still will outrank the newer administrative changes (like updating the ProgID version).

~~Can you please explain what a `ProgID` or `ProgIDVersion` is?~~ Edit: Ok I figured out, that `ProgID` is the name, MS uses and `ProgIDVersion` is your creation for the version suffix.

~~Who defines/changes the value?~~ Edit: We do. But _when_ should we increase it? What benefit does it have?

~~Why should the mime type differ between user or system-wide installation?~~ Edit: Ok the mime type is _not_ versioned, just the program association, right? So let's assume we had two hypothetical versions installed in parallel. How does a versioned ProgID help? The association still would point to only _one_ version. Would it automagically fallback to the other version, if the former gets uninstalled? If not, I really don't see the benefit.

overheadhunter commented on resources/innosetup/setup.iss:
```diff
@@ -61,25 +61,28 @@ Root: HKLM; Subkey: "SYSTEM\CurrentControlSet\Services\WebClient\Parameters"; Va
 ; Remove legacy AutoStart entry; see: https://github.com/cryptomator/cryptomator-win/issues/33
 Root: HKCU; Subkey: "SOFTWARE\Microsoft\Windows\CurrentVersion\Run\"; ValueType: none; ValueName: "Cryptomator"; Flags: dontcreatekey deletevalue
 
+; Remove legacy ProgID introduced in 1.5.12-beta1
+Root: HKCR; Subkey: "EncryptedCryptomatorVaultFile"; Flags: dontcreatekey deletekey
+
```
Would it harm to keep this value? If not, don't bother removing it, as we should not pile up cleanup code for pre-release versions.

overheadhunter replied:
It will get removed anyway during uninstall of the beta (note the `uninsdeletekey`):

https://github.com/cryptomator/cryptomator-win/blob/adc0e9139e98ef87441401603f79aab4981845ff/resources/innosetup/setup.iss#L65

JaniruTEC replied:
> Would it harm to keep this value? If not, don't bother removing it, as we should not pile up cleanup code for pre-release versions.

No, but it collides with Microsoft's specifications and could confuse users if the decide to snoop around in the registry. From a technical perspective, it doesn't matter.

> It will get removed anyway during uninstall of the beta (note the uninsdeletekey):

Correct, but since we don't call the uninstaller on update, this is far from a guarantee. 

Resolved by 7d3cb888975114ef02ccd2822780aabe0294dee0

overheadhunter replied:
> could confuse users

I hope so. Why should any sane person want to visit the registry, anyway?

> we don't call the uninstaller on update

But InnoSetup will append the install log, so it will get removed during uninstall eventually.

infeo commented on resources/innosetup/setup.iss:
```diff
@@ -57,15 +58,38 @@ Name: "webdav"; Description: "WebDAV system configuration"; Types: full; ExtraDi
+Root: HKA; Subkey: "Software\Classes\.c9s"; ValueType: string; ValueName: "PerceivedType"; ValueData: "system"
+
+Root: HKA; Subkey: "Software\Classes\Cryptomator.EncryptedData.{#ProgIDVersion}"; ValueType: string; ValueName: ""; ValueData: "Cryptomator Encrypted Data"; Flags: deletekey uninsdeletekey
+Root: HKA; Subkey: "Software\Classes\Cryptomator.EncryptedData.{#ProgIDVersion}\DefaultIcon"; ValueType: string; ValueName: ""; ValueData: "{app}\Cryptomator.exe,0"
```
When uninstalling Cryptomator, does this key need to be removed?

overheadhunter replied:
Same in Lines 78, 79. I guess it might get implicitly deleted via its parent key (line 90 and 77 respectively).

To quote [the InnoSetup docs](https://jrsoftware.org/ishelp/index.php?topic=registrysection):

> **uninsdeletekey**
> When the program is uninstalled, delete the entire key, including all values and subkeys in it.

overheadhunter replied:
Still, I agree that it _might_ be more dev friendly to explicitly add this flag to everything that is intended to be deleted during uninstall.

JaniruTEC replied:
Subkeys are in fact implicitly deleted by their parents. 

Therefore this is not the case:
> When uninstalling Cryptomator, does this key need to be removed?

--

> Still, I agree that it might be more dev friendly to explicitly add this flag to everything that is intended to be deleted during uninstall.

IMO the grouping does enough to highlight this fact and adding the flags manually, just adds additional complexity.

overheadhunter replied:
> IMO the grouping does enough to highlight this fact and adding the flags manually, just adds additional complexity.

Disagree. I don't see any complexity here, the line is already present anyway. It never hurts to be explicit about something that software is supposed to do. Instead, omitting this flag may confuse people (as proven by @infeo's initial comment) who don't know the recursive deletion behaviour of `uninsdeletekey` by heart.

JaniruTEC replied:
Alright, I will add the additional flags in the evening

JaniruTEC replied:
Resolved by 2447ddb

infeo committed:
Remove registrykey deletion code (job is done already with unsinstall)

infeo committed:
add comment to progID version

infeo committed:
Revert "Remove registrykey deletion code (job is done already with unsinstall)"

This reverts commit c2c40287ad45862721d4b4a71f6821f909c637a1.

infeo committed:
removed correct code part

JaniruTEC commented:
Generally speaking the Registry registers file extensions like this:
1. Create a key for the actual extension (e.g. `.cryptomator`) in HKCR (or HKLM/Software/Classes)
2. Create a matching ProgID (e.g. `Cryptomator.MasterkeyFile`)
3. Link the ProgID in the extension key
4. Define behavior for the extension in the *ProgID*

The benefit of this approach can be seen on the `.c9r` and `.c9s` extensions. Both reference the same ProgID, thus share the same Icon, description and behavior on doubleclick, but don't necessarily share a MIME Type or perceived type. 

> How does appending .1 to an arbitrary string protect a key from being deleted?

Those are different matters. Microsoft recommends to not delete the extension from the registry in case the user switched out the ProgID for a different one, that means: Changed the app that opens the file (another benefit of the dual approach)
ProgIDs should be deleted on uninstall (and install to clean them up and start fresh). If a extension contains a deleted/unknown ProgID, Windows falls back to asking the user.

> ~~Can you please explain what a ProgID or ProgIDVersion is?~~ Edit: Ok I figured out, that ProgID is the name, MS uses and ProgIDVersion is your creation for the version suffix.

ProgID: See above
Microsoft recommends using versioned ProgIDs to enable multi-version-install and allow better maintenance from a dev perspective. In fact I ran into a problem yesterday and using a versioned ProgID was the easiest solution. `ProgIDVersion` is just the current version suffix.

> ~~Who defines/changes the value?~~ Edit: We do. But when should we increase it? What benefit does it have?

As you said, we do. Increasing is usually is only necessary if the change something about the ProgID structure and need to make sure changes are reflected correctly. Also having a version ProgID allows better control by scripts on installation time.

> ~~Why should the mime type differ between user or system-wide installation?~~ Edit: Ok the mime type is not versioned, just the program association, right? So let's assume we had two hypothetical versions installed in parallel. How does a versioned ProgID help? The association still would point to only one version. Would it automagically fallback to the other version, if the former gets uninstalled? If not, I really don't see the benefit.

The ProgID is versioned, the file extension only references the versioned ProgID. 
The file extension wouldn't fallback to the other ProgID, but switching and updating becomes easier because a ProgID can be called "by name". See: https://web.archive.org/web/20210203223928/https://docs.microsoft.com/de-de/windows/win32/shell/fa-progids#using-versioned-programmatic-identifiers
